### ISSUE TITLE
TPC-H Query 1 Performance Issue

### ISSUE THREAD

**Issue by bajinsheng** [pingcap/tidb#56622]
## Bug Report

For query 1 in the TPC-H benchmark, I found that we can obtain a more efficient query plan with 50% reduced execution time by tuning the logic of this function: https://github.com/pingcap/tidb/blob/master/pkg/planner/core/task.go#L180

### 1. Minimal reproduce step (Required)

Query 1:
```sql
select l_returnflag, l_linestatus, sum(l_quantity) as sum_qty, sum(l_extendedprice) as sum_base_price, sum(l_extendedprice * (1 - l_discount)) as sum_disc_price, sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge, avg(l_quantity) as avg_qty, avg(l_extendedprice) as avg_price, avg(l_discount) as avg_disc, count(*) as count_order from LINEITEM where l_shipdate <= date '1998-12-01' - interval '108' day group by l_returnflag, l_linestatus order by l_returnflag, l_linestatus;
```

By applying this patch:
```patch
diff --git a/pkg/planner/core/task.go b/pkg/planner/core/task.go
index 14c49e3ae7..5aaa8236d5 100644
--- a/pkg/planner/core/task.go
+++ b/pkg/planner/core/task.go
@@ -178,14 +178,10 @@ func (p *PhysicalIndexJoin) Attach2Task(tasks ...base.Task) base.Task {
 
 // RowSize for cost model ver2 is simplified, always use this function to calculate row size.
 func getAvgRowSize(stats *property.StatsInfo, cols []*expression.Column) (size float64) {
-       if stats.HistColl != nil {
-               size = max(cardinality.GetAvgRowSizeDataInDiskByRows(stats.HistColl, cols), 0)
-       } else {
                // Estimate using just the type info.
                for _, col := range cols {
                        size += max(float64(chunk.EstimateTypeWidth(col.GetStaticType())), 0)
                }
-       }
        return
 }
```

We can observe a significant performance improvement, so perhaps there is somewhere we can optimize the code to generate the more efficient query plan in default for TiDB?

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)
Before:
```text
EXPLAIN ANALYZE FORMAT=verbose:
+----------------------------------+------------+---------------+---------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| id                               | estRows    | estCost       | actRows | task      | access object  | execution info                                                                                                                                                                                                                                                                                                                                               | operator info                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  | memory   | disk    |
+----------------------------------+------------+---------------+---------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| Sort_6                           | 1595523.03 | 2326377249.59 | 4       | root      |                | time:2.28s, loops:2, RU:8186.921973                                                                                                                                                                                                                                                                                                                          | tpch.lineitem.l_returnflag, tpch.lineitem.l_linestatus                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | 5.94 KB  | 0 Bytes |
| └─Projection_8                   | 1595523.03 | 593289278.81  | 4       | root      |                | time:2.28s, loops:4, Concurrency:5                                                                                                                                                                                                                                                                                                                           | tpch.lineitem.l_returnflag, tpch.lineitem.l_linestatus, Column#18, Column#19, Column#20, Column#21, Column#22, Column#23, Column#24, Column#25                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | 114.3 KB | N/A     |
|   └─HashAgg_14                   | 1595523.03 | 591696946.83  | 4       | root      |                | time:2.28s, loops:4, partial_worker:{wall_time:2.277190096s, concurrency:5, task_num:1, tot_wait:11.385495695s, tot_exec:165.015µs, tot_time:11.38567701s, max:2.277144004s, p95:2.277144004s}, final_worker:{wall_time:2.277295269s, concurrency:5, task_num:5, tot_wait:8µs, tot_exec:240ns, tot_time:11.386134553s, max:2.277256667s, p95:2.277256667s}   | group by:tpch.lineitem.l_linestatus, tpch.lineitem.l_returnflag, funcs:sum(Column#26)->Column#18, funcs:sum(Column#27)->Column#19, funcs:sum(Column#28)->Column#20, funcs:sum(Column#29)->Column#21, funcs:avg(Column#30, Column#31)->Column#22, funcs:avg(Column#32, Column#33)->Column#23, funcs:avg(Column#34, Column#35)->Column#24, funcs:count(Column#36)->Column#25, funcs:firstrow(tpch.lineitem.l_returnflag)->tpch.lineitem.l_returnflag, funcs:firstrow(tpch.lineitem.l_linestatus)->tpch.lineitem.l_linestatus                                                                                                                                                                                                                     | 124.4 KB | 0 Bytes |
|     └─TableReader_15             | 1595523.03 | 318095799.20  | 136     | root      |                | time:2.28s, loops:2, cop_task: {num: 34, max: 849.9ms, min: 557.9ms, avg: 721.5ms, p95: 831ms, tot_proc: 24.5s, copr_cache_hit_ratio: 0.00, build_task_duration: 52.1µs, max_distsql_concurrency: 15}, rpc_info:{Cop:{num_rpc:34, total_time:24.5s}}                                                                                                         | data:HashAgg_9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | 3.07 KB  | N/A     |
|       └─HashAgg_9                | 1595523.03 | 2787499834.04 | 136     | cop[tikv] |                | tikv_task:{proc max:849.9ms, min:557.8ms, avg: 721.5ms, p80:742.2ms, p95:830.9ms, iters:0, tasks:34}, time_detail: {total_process_time: 24.5s}                                                                                                                                                                                                               | group by:tpch.lineitem.l_linestatus, tpch.lineitem.l_returnflag, funcs:sum(tpch.lineitem.l_quantity)->Column#26, funcs:sum(tpch.lineitem.l_extendedprice)->Column#27, funcs:sum(mul(tpch.lineitem.l_extendedprice, minus(1, tpch.lineitem.l_discount)))->Column#28, funcs:sum(mul(mul(tpch.lineitem.l_extendedprice, minus(1, tpch.lineitem.l_discount)), plus(1, tpch.lineitem.l_tax)))->Column#29, funcs:count(tpch.lineitem.l_quantity)->Column#30, funcs:sum(tpch.lineitem.l_quantity)->Column#31, funcs:count(tpch.lineitem.l_extendedprice)->Column#32, funcs:sum(tpch.lineitem.l_extendedprice)->Column#33, funcs:count(tpch.lineitem.l_discount)->Column#34, funcs:sum(tpch.lineitem.l_discount)->Column#35, funcs:count(1)->Column#36 | N/A      | N/A     |
|         └─Selection_13           | 1994403.78 | 2440632428.51 | 5879628 | cop[tikv] |                | tikv_task:{proc max:849.9ms, min:557.8ms, avg: 721.5ms, p80:742.2ms, p95:830.9ms, iters:0, tasks:34}                                                                                                                                                                                                                                                         | le(tpch.lineitem.l_shipdate, 1998-08-15)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | N/A      | N/A     |
|           └─TableFullScan_12     | 6001215.00 | 2141171800.01 | 5879628 | cop[tikv] | table:LINEITEM | tikv_task:{proc max:849.9ms, min:557.8ms, avg: 721.5ms, p80:742.2ms, p95:830.9ms, iters:0, tasks:34}                                                                                                                                                                                                                                                         | keep order:false, stats:pseudo                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | N/A      | N/A     |
+----------------------------------+------------+---------------+---------+-----------+----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
```
```bash
mysql> select l_returnflag, l_linestatus, sum(l_quantity) as sum_qty, sum(l_extendedprice) as sum_base_price, sum(l_extendedprice * (1 - l_discount)) as sum_disc_price, sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge, avg(l_quantity) as avg_qty, avg(l_extendedprice) as avg_price
, avg(l_discount) as avg_disc, count(*) as count_order from LINEITEM where l_shipdate <= date '1998-12-01' - interval '108' day group by l_returnflag, l_linestatus order by l_returnflag, l_linestatus;
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
| l_returnflag | l_linestatus | sum_qty     | sum_base_price  | sum_disc_price    | sum_charge          | avg_qty   | avg_price    | avg_disc | count_order |
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
| A            | F            | 37734107.00 |  56586554400.73 |  53758257134.8700 |  55909065222.827692 | 25.522006 | 38273.129735 | 0.049985 |     1478493 |
| N            | F            |   991417.00 |   1487504710.38 |   1413082168.0541 |   1469649223.194375 | 25.516472 | 38284.467761 | 0.050093 |       38854 |
| N            | O            | 73533166.00 | 110287596362.18 | 104774847005.9449 | 108969626230.358561 | 25.502145 | 38249.003129 | 0.049996 |     2883411 |
| R            | F            | 37719753.00 |  56568041380.90 |  53741292684.6040 |  55889619119.831932 | 25.505794 | 38250.854626 | 0.050009 |     1478870 |
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
4 rows in set (2.24 sec)
```

After:
```text
EXPLAIN ANALYZE FORMAT=verbose:
+----------------------------------+------------+---------------+---------+-----------+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| id                               | estRows    | estCost       | actRows | task      | access object  | execution info                                                                                                                                                                                                                                                                                                                                                  | operator info                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | memory   | disk    |
+----------------------------------+------------+---------------+---------+-----------+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| Sort_6                           | 1595523.03 | 2303520913.95 | 4       | root      |                | time:969.1ms, loops:2, RU:19320.328530                                                                                                                                                                                                                                                                                                                          | tpch.lineitem.l_returnflag, tpch.lineitem.l_linestatus                                                                                                                                                                                                                                                                                                                                                                                                                                   | 5.94 KB  | 0 Bytes |
| └─Projection_8                   | 1595523.03 | 570432943.18  | 4       | root      |                | time:969ms, loops:4, Concurrency:5                                                                                                                                                                                                                                                                                                                              | tpch.lineitem.l_returnflag, tpch.lineitem.l_linestatus, Column#18, Column#19, Column#20, Column#21, Column#22, Column#23, Column#24, Column#25                                                                                                                                                                                                                                                                                                                                           | 114.3 KB | N/A     |
|   └─HashAgg_11                   | 1595523.03 | 568840611.20  | 4       | root      |                | time:968.9ms, loops:4, partial_worker:{wall_time:968.861723ms, concurrency:5, task_num:5723, tot_wait:2.0839277s, tot_exec:2.758085007s, tot_time:4.84383593s, max:968.774ms, p95:968.774ms}, final_worker:{wall_time:968.929365ms, concurrency:5, task_num:17, tot_wait:12.9µs, tot_exec:25.49µs, tot_time:4.844043088s, max:968.832422ms, p95:968.832422ms}   | group by:Column#46, Column#47, funcs:sum(Column#39)->Column#18, funcs:sum(Column#40)->Column#19, funcs:sum(Column#41)->Column#20, funcs:sum(Column#42)->Column#21, funcs:avg(Column#43)->Column#22, funcs:avg(Column#44)->Column#23, funcs:avg(Column#45)->Column#24, funcs:count(1)->Column#25, funcs:firstrow(Column#46)->tpch.lineitem.l_returnflag, funcs:firstrow(Column#47)->tpch.lineitem.l_linestatus                                                                            | 6.41 MB  | 0 Bytes |
|     └─Projection_20              | 1994403.78 | 284610647.58  | 5879628 | root      |                | time:847.2ms, loops:5724, Concurrency:5                                                                                                                                                                                                                                                                                                                         | tpch.lineitem.l_quantity->Column#39, tpch.lineitem.l_extendedprice->Column#40, mul(tpch.lineitem.l_extendedprice, minus(1, tpch.lineitem.l_discount))->Column#41, mul(mul(tpch.lineitem.l_extendedprice, minus(1, tpch.lineitem.l_discount)), plus(1, tpch.lineitem.l_tax))->Column#42, tpch.lineitem.l_quantity->Column#43, tpch.lineitem.l_extendedprice->Column#44, tpch.lineitem.l_discount->Column#45, tpch.lineitem.l_returnflag->Column#46, tpch.lineitem.l_linestatus->Column#47 | 3.36 MB  | N/A     |
|       └─TableReader_18           | 1994403.78 | 243409057.55  | 5879628 | root      |                | time:628.4ms, loops:5724, cop_task: {num: 407, max: 65.2ms, min: 236.6µs, avg: 17.7ms, p95: 59.4ms, tot_proc: 6.78s, copr_cache_hit_ratio: 0.00, build_task_duration: 53.5µs, max_distsql_concurrency: 15}, rpc_info:{Cop:{num_rpc:407, total_time:7.18s}}                                                                                                      | data:Selection_17                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | 80.7 MB  | N/A     |
|         └─Selection_17           | 1994403.78 | 2308503235.20 | 5879628 | cop[tikv] |                | tikv_task:{proc max:63.5ms, min:219.1µs, avg: 17.1ms, p80:36.5ms, p95:57.4ms, iters:0, tasks:407}, time_detail: {total_process_time: 6.78s}                                                                                                                                                                                                                     | le(tpch.lineitem.l_shipdate, 1998-08-15)                                                                                                                                                                                                                                                                                                                                                                                                                                                 | N/A      | N/A     |
|           └─TableFullScan_16     | 6001215.00 | 2009042606.70 | 6001215 | cop[tikv] | table:LINEITEM | tikv_task:{proc max:63.5ms, min:219.1µs, avg: 17.1ms, p80:36.5ms, p95:57.4ms, iters:0, tasks:407}                                                                                                                                                                                                                                                               | keep order:false, stats:pseudo                                                                                                                                                                                                                                                                                                                                                                                                                                                           | N/A      | N/A     |
+----------------------------------+------------+---------------+---------+-----------+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
```
```bash
mysql> select l_returnflag, l_linestatus, sum(l_quantity) as sum_qty, sum(l_extendedprice) as sum_base_price, sum(l_extendedprice * (1 - l_discount)) as sum_disc_price, sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge, avg(l_quantity) as avg_qty, avg(l_extendedprice) as avg_price, avg(l_discount) as avg_disc, count(*) as count_order from LINEITEM where l_shipdate <= date '1998-12-01' - interval '108' day group by l_returnflag, l_linestatus order by l_returnflag, l_linestatus;
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
| l_returnflag | l_linestatus | sum_qty     | sum_base_price  | sum_disc_price    | sum_charge          | avg_qty   | avg_price    | avg_disc | count_order |
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
| A            | F            | 37734107.00 |  56586554400.73 |  53758257134.8700 |  55909065222.827692 | 25.522006 | 38273.129735 | 0.049985 |     1478493 |
| N            | F            |   991417.00 |   1487504710.38 |   1413082168.0541 |   1469649223.194375 | 25.516472 | 38284.467761 | 0.050093 |       38854 |
| N            | O            | 73533166.00 | 110287596362.18 | 104774847005.9449 | 108969626230.358561 | 25.502145 | 38249.003129 | 0.049996 |     2883411 |
| R            | F            | 37719753.00 |  56568041380.90 |  53741292684.6040 |  55889619119.831932 | 25.505794 | 38250.854626 | 0.050009 |     1478870 |
+--------------+--------------+-------------+-----------------+-------------------+---------------------+-----------+--------------+----------+-------------+
4 rows in set (0.98 sec)
```

### 4. What is your TiDB version? (Required)
```
| Release Version: v8.4.0-alpha-382-g3dfc47f2cc-dirty
Edition: Community
Git Commit Hash: 3dfc47f2cc289a4f39843393f98da926f0c7ba1e
Git Branch: HEAD
UTC Build Time: 2024-10-13 19:19:05
GoVersion: go1.23.2
Race Enabled: false
Check Table Before Drop: false
Store: unistore |
```

I generated 1 GB of data for TPC-H and shared my data directory in [Google drive](https://drive.google.com/file/d/1-CGqmtV9Fwn3XslmT3vUVGdVGh1q4Zid/view?usp=sharing) for your analysis.

**Comment by winoros** [comment]
first of all, how to you measure the perf? by running multiple times then calculating the avg?
the result is strange. please also paste your cluster topology and the detailed info of each node.

**Comment by winoros** [comment]
oh, if you are going to eval perf. please use tikv/tiflash, instead of the embedded unistore.
we don't guarantee any perf index of the unistore. it's only used to do functionality tests.

and use the scale factor at least 10 to test the ap perf.

**Comment by bajinsheng** [comment]
>how to you measure the perf?

I found the execution time would decrease over executions, so I executed the test case around 10 times and got the final execution time, which seems more stable.

> if you are going to eval perf. please use tikv/tiflash

I see. With the help from @hawkingrei, I reconfigured my testing environment with tiup playground. I used this `tiup playground nightly --mode tikv-slim` together with a standalone TiDB instance. I also disabled the coprocessor cache. For the topology, I used the default configuration, which seems 1 pd, 1 tikv, and 1 tidb? All of them are in the same machine.

I rerun this test case, and both executions took a similar time, around 1 second, so the performance gap disappeared for this case. However, for the other case https://github.com/pingcap/tidb/issues/56623 I reported, the performance gap still exists.

Note that I still use the scale factor of 1 for TPC-H, and will try factor 10 later as it requires much longer time.
